### PR TITLE
Enhanced stash status key

### PR DESCRIPTION
**Pull request by ahammond**
Reopening PR #1904. Addressed all comments except the one about removing inlineCallbacks for the send method (which I tried, but suggestion didn't work and I'm only a beginner with twisted).

### CODE REVIEW

**Comment by tardyp** [comment]
Hello,
thanks for your share.
I dont think we want new features in the buildbot eight branch, now that nine is nearly finished. Would you like to port those changes to the nine branch?

**Comment by ahammond** [comment]
Happy to port it to 9. It'd give me an excuse to test 9 for deployment here. That said, I'd love to get this merged to 8 since it's what we're using in production and have been for many months now.

**Comment by tardyp** [comment]
lets tag it nine-first then. We will see what @sa2ajj decides

**Comment by sa2ajj** [comment]
First, general comment: I'm willing to take this change (thank you, @ahammond).  However I'd like to release **last** 0.8.x version after 0.9 is released.  In essence, this means that things that are not in `master` are not going to be maintained after that.

File: master/buildbot/status/status_stash.py [hunk 1]
```diff
@@ -71,45 +71,76 @@ class StashStatusPush(StatusReceiverMultiService,
                       BuildSetSummaryNotifierMixin):
     implements(IStatusReceiver)
 
-    def __init__(self, base_url, user, password):
+    def __init__(self, base_url, user, password,
+                 key_format='%(builderName)s',
```
Comment (sa2ajj) [inline master/buildbot/status/status_stash.py]: why do you want to have such a complicated implementation?  you use `builder.render` for various things below, why not use it to **just** render the `key_format` (and `name_format`) parameter?

File: master/docs/manual/cfg-statustargets.rst [hunk 2]
```diff
@@ -1639,13 +1641,18 @@ Specifically, it follows the `Updating build status for commits <https://develop
 
 It uses the standard Python Twisted Agent to make REST requests to the stash server.
 It uses HTTP Basic AUTH.
-As a result, we recommend you use https in your base_url rather than http.
+As a result, we recommend you either connect over a secured network or use https in your base_url rather than http.
```
Comment (sa2ajj) [inline master/docs/manual/cfg-statustargets.rst]: nice touch :smile:

File: master/docs/manual/cfg-statustargets.rst [hunk 3]
```diff
@@ -1639,13 +1641,18 @@ Specifically, it follows the `Updating build status for commits <https://develop
...
 If you use https, it requires `pyOpenSSL`.
 
-Configuration requires exactly 3 parameters:
+Configuration requires 3 parameters:
 `base_url` is the base url of the stash host, up to and optionally including the first / of the path.
 `user` is the stash user to post as
 `password` is the stash user's password
+There are also two optional parameters:
+`key_format` is a python format string used to define the aggregation key for builds in stash.
```
Comment (sa2ajj) [inline master/docs/manual/cfg-statustargets.rst]: I know that `python` is not spelled consistently over the documentation, however I'd prefer that new docs has `Python` all over... :smile:

Comment (sa2ajj) [inline master/docs/manual/cfg-statustargets.rst]: Did you consider formatting these parameters as

```
`key_format`
    Python format string ....
```

**Comment by sa2ajj** [comment]
Otherwise LGTM (the tests failurs **I** need to look into).

File: master/buildbot/status/status_stash.py (hunk 1 above)
Comment (ahammond) [inline master/buildbot/status/status_stash.py]: Now... that's a good idea!

**Comment by ahammond** [comment]
Updated per @sa2ajj 's comments and pushed.

**Comment by ahammond** [comment]
Is there anything left for me to do with this?

**Comment by sa2ajj** [comment]
@ahammond what are your plans regarding nine (aka master)? :)

**Comment by ahammond** [comment]
I don't have any plans yet. If there's a porting guide, I'm happy to merge
forward the code I've already written. Otherwise, I'd really like some
pointers to get started.

**Comment by djmitche** [comment]
The new stash support is in
  https://github.com/buildbot/buildbot/blob/master/master/buildbot/reporters/stash.py
if you would like to take a look at that.

We are not expecting to make another 0.8.x release, so we are not accepting new functionality in the eight branch that is not also present in the master branch.

**Comment by sa2ajj** [comment]
based on our discussion [yesterday](https://lists.buildbot.net/pipermail/devel/2016-June/012173.html) I will make a 0.8.x release in the coming days.

If you'd like your changes to be maintained afterwards, please consider submitting a PR against master.

(I'll keep this PR open until the next 0.8.x release is about to be made.)

**Pull request merged**